apps/exam/models.py

```diff
@@ -66,6 +68,30 @@ def time_left_seconds(self):
     time_left = end_time - timezone.now()
     return max(int(time_left.total_seconds()), 0)
 
+  def calculate_grade(self):
+    """Calculate the grade for an attempt."""
+
+    correct_answer_count = 0
+    grade = 0
+    answer_attempts = AnswerAttempt.objects.filter(attempt=self.id)
+    question_ids = [a.question.id for a in answer_attempts]
+
+    if answer_attempts:
+      for question_id in question_ids:
```

Comment:
You could use a loop over `AnswerAttempt.objects.filter(question_id__in=question_ids, attempt=self.id)`

Also take a look at .prefetch_related() - https://docs.djangoproject.com/en/4.0/ref/models/querysets/#prefetch-related for answers and .select_related() - https://docs.djangoproject.com/en/4.0/ref/models/querysets/#select-related for questions